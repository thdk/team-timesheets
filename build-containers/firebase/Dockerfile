FROM node:16 AS node_base

RUN echo "NODE Version:" && node --version
RUN echo "NPM Version:" && npm --version

FROM openjdk:8-alpine
COPY --from=node_base . .

# install Firebase CLI
RUN npm install -g firebase-tools
RUN firebase setup:emulators:firestore

ENTRYPOINT ["/usr/local/bin/firebase"]