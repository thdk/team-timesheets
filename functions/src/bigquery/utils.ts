// Imports the Google Cloud client library
import { BigQuery } from '@google-cloud/bigquery';
import { BigQueryTableSchemes } from './config';

export type BigQueryField = { name: string, type: string, mode?: string };
export type BigQueryConfig = {
    dataSetId: string;
    tableId: string;
    tableIdPrefix: string,
    schemes: BigQueryTableSchemes,
    timePartitioningField?: string,
};

function validateTableSchemaAsync(table: any, schema: BigQueryField[]) {
    return table.getMetadata().then(([metadata]) => {
        if (metadata.schema) {
            console.log(`Current schema of table ${table.id} is ${JSON.stringify(metadata.schema.fields)}`);
        } else {
            console.log("Table has no schema yet.");
        }

        console.log(`Required schema of table ${table.id} is ${JSON.stringify(schema)}`);

        if (!isEqualSchema(metadata.schema.fields, schema)) {
            console.log(`Updating schema of table ${table.id}...`);
            return table.setMetadata(Object.assign(metadata, { schema }))
                .then(() => {
                    console.log("Schema updated.");
                    return table;
                }, () => {
                    console.log("Couldn't update schema of table: " + table.id);
                });
        }
        else {
            console.log(`Schema of table ${table.id}... is already up to date.`);
            return new Promise(resolve => resolve(table));
        }
    });
}

function isEqualSchema(schema1: BigQueryField[], schema2: BigQueryField[]) {
    console.log(`Comparing fields length... ${schema1.length} <> ${schema2.length}`);
    if (schema1.length !== schema2.length) return false;

    return !schema1.some((field1, index) => {

        const field2 = schema2[index];
        console.log(`Comparing field...`);
        console.log(`${JSON.stringify(field1)}`);
        console.log(`${JSON.stringify(field2)}`);

        return field1.name !== field2.name ||
            field1.type !== field2.type ||
            field1.mode !== field2.mode;
    });
}

type InsertOptions = {
    schema: BigQueryField[];
    location: string;
    timePartitioning?: {
        type: 'DAY';
        field: string;
    };
}
export function insertRowsAsync<T>(options: BigQueryConfig, rows: ReadonlyArray<T>, bigqueryClient?: BigQuery): Promise<void> {
    const { dataSetId, tableId, tableIdPrefix = "", schemes, timePartitioningField } = options;

    console.log(`Inserting ${rows.length} rows into ${tableId}`);

    let insertOptions: InsertOptions = {
        schema: schemes[tableId],
        location: 'US'
    };

    if (timePartitioningField) {
        insertOptions = {
            ...insertOptions,
            timePartitioning: {
                type: 'DAY',
                field: timePartitioningField,
            },
        }
    }

    if (!rows.length) return new Promise(resolve => resolve());

    const bigquery = bigqueryClient || new BigQuery();

    // TODO: create dataset only once since insertRows can be called multiple times simultaneously!
    return bigquery
        .dataset(dataSetId)
        .get({ autoCreate: true })
        .then(([dataset]) => {
            console.log("dataset created");
            return dataset.table(`${tableIdPrefix}${tableId}`)
                .get({ autoCreate: true, ...insertOptions })
                .then(([table]) => validateTableSchemaAsync(table, insertOptions.schema))
                .then(table => {
                    const first10kRows = rows.slice(0, 10000);
                    return table.insert(first10kRows)
                        .then(() => {
                            console.log(`Inserted ${first10kRows.length} rows of total ${rows.length} into ${tableId}`);
                            if (rows.length > 10000) {
                                console.log("Inserting next batch...")
                                return insertRowsAsync(options, rows.slice(10000));
                            }

                            console.log(`Finished inserting into ${tableId}`);
                            return true;
                        }, (error: { errors: any, name: string, response: any, message: string }) => {
                            error.errors && error.errors.forEach(e => {
                                console.log({
                                    error: error.name,
                                    message: error.message,
                                    errors: e.errors,
                                    row: e.row
                                });
                            });
                        }, (e) => {
                            console.log(e);
                            console.log("Table could not be created");
                        })
                })
        }, e => {
            console.log(e);
            console.log("Dataset could not be created");
        });

}