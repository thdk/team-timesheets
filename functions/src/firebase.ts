import * as admin from "firebase-admin";

admin.initializeApp({ credential: admin.credential.applicationDefault() });

export const db = admin.firestore();
