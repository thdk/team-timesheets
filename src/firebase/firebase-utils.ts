import type firebase from "firebase";

/**
 * Resolves with firbase.User if user is logged in
 * Rejects if no user is logged in
 */
export const getLoggedInUserAsync = (auth: firebase.auth.Auth) => {
    return new Promise<firebase.User>((resolve, reject) => {
        const unsubscribe = auth.onAuthStateChanged(user => {
            unsubscribe();
            if (user) resolve(user);
            else reject("Not authenticated");
        });
    });
}