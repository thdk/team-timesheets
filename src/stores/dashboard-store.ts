import { observable, reaction, computed, action } from "mobx";
import { ICollection, Collection, RealtimeMode } from "firestorable";
import type firebase from "firebase";

import moment from 'moment';

import { IRootStore } from "./root-store";
import { IGroupedRegistrations } from "./registration-store/registration-store";
import * as deserializer from '../../common/serialization/deserializer';
import * as serializer from '../../common/serialization/serializer';
import { TimePeriod } from "../components/time-period-select";
import { IRegistration, IRegistrationData } from "../../common/dist";

export interface IDashboardStore {
    readonly setProjectFilter: (projectId: string | undefined) => void;
    readonly setUserFilter: (userId: string | undefined) => void;
    readonly setTaskFilter: (taskId: string | undefined) => void;
    readonly setTimePeriodFilter: (timePeriod: TimePeriod) => void;

    readonly registrationsGroupedByUser: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByTask: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByProject: IGroupedRegistrations<string>[];

    readonly projectFilterValue?: string;
    readonly taskFilterValue?: string;
    readonly userFilterValue?: string;
    readonly timePeriodFilterValue?: TimePeriod;

}

export class DashboardStore implements IDashboardStore {
    private readonly registrationsField: ICollection<IRegistration, IRegistrationData>;

    @observable.ref private userFilterValueField: string | undefined = undefined;
    @observable.ref private taskFilterValueField: string | undefined = undefined;
    @observable.ref private projectFilterValueField: string | undefined = undefined;
    @observable.ref private timePeriodFilterField: TimePeriod | undefined = undefined;

    constructor(
        rootStore: IRootStore,
        {
            firestore,
        }: {
            firestore: firebase.firestore.Firestore,
        },
    ) {
        this.registrationsField = new Collection<IRegistration, IRegistrationData>(
            firestore,
            "registrations",
            {
                realtimeMode: RealtimeMode.on,
                deserialize: deserializer.convertRegistration,
                serialize: serializer.convertRegistration,
                name: "Dashboard registrations",
            },
            {
                // logger: console.log
            },
        );

        const updateRegistrationQuery = () => {
            if (!rootStore.user.divisionUser || !this.timePeriodFilterField) {
                this.registrationsField.query = null;
            } else {
                this.registrationsField.query = ref =>
                    this.userIdFilter(
                        this.projectIdFilter(
                            this.endDateFilter(
                                this.startDateFilter(
                                    ref.where("deleted", "==", false)
                                )
                            )
                        )
                    );
            }
        };

        // update the query of the registration collection each time...
        // -- a filter changes (userId, startDate, endDate)
        // TODO: add a Filter button in UI to group all filter changes into one new search query
        reaction(() => this.timePeriodFilterField, updateRegistrationQuery);
        reaction(() => this.userFilterValue, updateRegistrationQuery);
        reaction(() => this.projectFilterValue, updateRegistrationQuery);
        reaction(() => rootStore.user.divisionUser, updateRegistrationQuery);
    }

    @computed
    private get startDateFilter() {
        const startDate = this.startDate;
        return (query: firebase.firestore.Query) => startDate ? query.where("date", ">=", startDate) : query;
    }

    @computed
    private get endDateFilter() {
        const endDate = this.endDate;
        return (query: firebase.firestore.Query) => endDate ? query.where("date", "<=", endDate) : query;
    }

    @computed
    private get userIdFilter() {
        const userId = this.userFilterValue;
        return (query: firebase.firestore.Query) => userId ? query.where("userId", "==", userId) : query;
    }

    @computed
    private get projectIdFilter() {
        return (query: firebase.firestore.Query) => this.projectFilterValue
            ? query.where("project", "==", this.projectFilterValue)
            : query;
    }

    @computed
    get registrationsGroupedByUser() {
        // no need to group when userId is included in the filter
        if (this.registrations.length === 0) return [];

        const registrationsPerUser = this.createRegistrationsGroup(reg => reg.userId);

        return registrationsPerUser;
    }

    @computed
    get registrationsGroupedByProject() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.project || "")
            .sort((a, b) => b.totalTime - a.totalTime);;
    }

    @computed
    get registrationsGroupedByTask() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.task || "")
            .sort((a, b) => b.totalTime - a.totalTime);
    }

    @computed
    private get startDate() {
        const today = moment(new Date());
        switch (this.timePeriodFilterField) {
            case TimePeriod.ThisWeek:
                return today.clone().startOf('week').toDate();
            case TimePeriod.ThisMonth:
                return today.clone().startOf('month').toDate();
            case TimePeriod.LastMonth:
                return today.clone().subtract(1, 'months').startOf('month').toDate();
            case TimePeriod.ThisYear:
                return today.clone().startOf('year').toDate();
            case TimePeriod.LastYear:
                return today.clone().subtract(1, 'years').startOf('year').toDate();
            default:
                return undefined;
        }
    }

    @computed
    private get endDate() {
        const today = moment(new Date());
        switch (this.timePeriodFilterField) {
            case TimePeriod.ThisWeek:
                return today.clone().endOf('week').toDate();
            case TimePeriod.ThisMonth:
                return today.clone().endOf('month').toDate();
            case TimePeriod.LastMonth:
                return today.clone().subtract(1, 'months').endOf('month').toDate();
            case TimePeriod.ThisYear:
                return today.clone().endOf('year').toDate();
            case TimePeriod.LastYear:
                return today.clone().subtract(1, 'years').endOf('year').toDate();
            default:
                return undefined;
        }
    }

    public get taskFilterValue() {
        return this.taskFilterValueField;
    }

    public get userFilterValue() {
        return this.userFilterValueField;
    }

    public get projectFilterValue() {
        return this.projectFilterValueField;
    }

    public get timePeriodFilterValue() {
        return this.timePeriodFilterField;
    }

    @computed
    private get registrations() {
        return this.registrationsField.docs
            .filter(doc => doc.data!.isPersisted) // don't display drafts (TODO: move to collection.ts?)
    }

    @action
    public setProjectFilter(projectId: string | undefined) {
        this.projectFilterValueField = projectId;
    }

    @action
    public setTaskFilter(taskId: string | undefined) {
        this.taskFilterValueField = taskId;
    }

    @action
    public setUserFilter(userId: string | undefined) {
        this.userFilterValueField = userId;
    }

    @action setTimePeriodFilter(timePeriod: TimePeriod) {
        this.timePeriodFilterField = timePeriod;
    }

    private createRegistrationsGroup = <T>(keySelector: (reg: IRegistration) => T): IGroupedRegistrations<T>[] => {
        const groupedMap = new Map<T, IGroupedRegistrations<T>>();

        this.registrations
            .forEach(c => {
                const key = keySelector(c.data!);
                const group = groupedMap.get(key);
                if (group) {
                    group.totalTime += c.data!.time || 0;
                    group.registrations.push(c);
                }
                else {
                    groupedMap.set(key, {
                        groupKey: keySelector(c.data!),
                        registrations: [c],
                        totalTime: c.data!.time || 0,
                        isCollapsed: true
                    });
                }
            });

        return Array.from(groupedMap.values());
    }
}