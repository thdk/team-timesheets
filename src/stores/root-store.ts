import { IRegistrationsStore, RegistrationStore } from "./registration-store";
import { IConfigStore, ConfigStore } from "./config-store";
import { IUserStore, UserStore } from "./user-store";
import { IViewStore, ViewStore } from "./view-store";
import { RouterStore } from "mobx-router";
import { IReportStore, ReportStore } from "./report-store";
import { DashboardStore, IDashboardStore } from "./dashboard-store";
import { IProjectStore, ProjectStore } from "./project-store";
import { FavoriteStore } from "./favorite-store";

export interface IRootStore {
    readonly user: IUserStore;
    readonly view: IViewStore;
    readonly router: RouterStore;
    readonly timesheets: IRegistrationsStore;
    readonly reports: IReportStore;
    readonly config: IConfigStore;
    readonly projects: IProjectStore;
    readonly favorites: FavoriteStore;
    readonly dashboard: IDashboardStore;
}

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router = new RouterStore();
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: IProjectStore;
    public readonly favorites: FavoriteStore;

    constructor() {
        this.user = new UserStore(this);
        this.view = new ViewStore(this);
        this.config = new ConfigStore(this);
        this.timesheets = new RegistrationStore(this);
        this.reports = new ReportStore(this);
        this.dashboard = new DashboardStore(this);
        this.projects = new ProjectStore(this);
        this.favorites = new FavoriteStore(this);
    }
};
