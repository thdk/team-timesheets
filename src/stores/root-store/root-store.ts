import { RouterStore } from "mobx-router";

import { IRegistrationsStore, RegistrationStore } from "../registration-store/registration-store";
import { IConfigStore, ConfigStore } from "../config-store";
import { IUserStore, UserStore } from "../user-store";
import { IViewStore, ViewStore } from "../view-store";
import { IReportStore, ReportStore } from "../report-store/report-store";
import { DashboardStore, IDashboardStore } from "../dashboard-store";
import { ProjectStore } from "../project-store";
import { FavoriteStore } from "../favorite-store";
import { DivisionStore  } from "../division-store/division-store";

export interface IRootStore extends Store { };

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router: RouterStore<IRootStore>;
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: ProjectStore;
    public readonly favorites: FavoriteStore;
    public readonly divisions: DivisionStore;

    constructor({
        auth,
        firestore,
        httpsCallable,
    }: {
        firestore: firebase.firestore.Firestore,
        httpsCallable?: (name: string) => firebase.functions.HttpsCallable,
        auth?: firebase.auth.Auth,
    }) {
        this.user = new UserStore(
            this,
            {
                auth,
                firestore,
            }
        );

        this.view = new ViewStore();
        this.config = new ConfigStore(
            this,
            {
                firestore,
            },
        );
        this.timesheets = new RegistrationStore(
            this,
            {
                firestore,
            },
        );
        this.reports = new ReportStore(
            this,
            {
                firestore,
            },
        );
        this.dashboard = new DashboardStore(
            this,
            {
                firestore,
            },
        );
        this.projects = new ProjectStore(
            this,
            {
                firestore,
            },
        );
        this.favorites = new FavoriteStore(
            this,
            {
                firestore,
            },
        );
        this.divisions = new DivisionStore(
            this,
            {
                firestore,
                httpsCallable,
            },
        );
        this.router = new RouterStore<IRootStore>(this);
    }

    public dispose() {
        this.config.dispose();
        // this.dashboard.dispose();
        // this.favorites.dispose();
        // this.projects.dispose();
        // this.reports.dispose();
        this.timesheets.dispose();
        this.user.dispose();
        // this.view.dispose();
    }
};
