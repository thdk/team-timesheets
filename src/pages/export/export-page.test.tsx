import React from "react";
import { initTestFirestore, deleteFirebaseAppsAsync } from "../../__tests__/utils/firebase";
import { Store } from "../../stores/root-store";
import { render, waitFor } from "@testing-library/react";
import { ExportPage } from ".";

const {
    firestore,
    clearFirestoreDataAsync,
} = initTestFirestore("export-page-test",
    [
        "registrations",
    ]);

const store = new Store({
    firestore,
});

jest.mock("../../contexts/store-context", () => ({
    useStore: () => store,
}));

jest.mock("../../contexts/user-context", () => ({
    useUserStore: jest.fn().mockReturnValue({
        authenticatedUser: {},
    }),
}));

jest.mock("../../rules");
jest.mock("../../routes/registrations/detail");


beforeAll(clearFirestoreDataAsync);
afterAll(() => {
    store.dispose();
    return Promise.all([
        deleteFirebaseAppsAsync(),
    ])
});

const registrations = [
    {
        date: new Date(2020, 2, 1),
        userId: "user-1",
        project: "project-1",
        task: "task-1",
        time: 4.5,
        isPersisted: true,
        description: "Foobar 1",
    },
    {
        date: new Date(2020, 2, 2),
        userId: "user-1",
        project: "project-1",
        task: "task-1",
        time: 4.5,
        isPersisted: true,
        description: "Foobar 2",
    },
    {
        date: new Date(2020, 2, 24),
        userId: "user-1",
        project: "project-1",
        task: "task-1",
        time: 2,
        isPersisted: true,
        description: "Foobar 3",
    },
    {
        date: new Date(2020, 2, 24),
        userId: "user-1",
        project: "project-2",
        task: "task-1",
        time: 2,
        isPersisted: true,
        description: "Foobar 4",
    },
    {
        date: new Date(2020, 2, 24),
        userId: "user-1",
        project: "project-2",
        task: "task-2",
        time: 6,
        isPersisted: true,
        description: "Foobar 5",
    },
    {
        date: new Date(2020, 3, 1),
        userId: "user-1",
        project: "project-1",
        task: "task-1",
        time: 2.5,
        isPersisted: true,
        description: "Another month"
    },
];

describe("Export Page", () => {
    it("should render without registrations", () => {
        const { asFragment } = render(<ExportPage />);

        expect(asFragment()).toMatchSnapshot();
    });

    it("should display report for the specified month", async () => {
        store.view.setViewDate({
            year: 2020,
            month: 3,
        });

        store.user.setUser({ uid: "user-1" } as firebase.User);

        const {
            asFragment,
            getByText,
            container,
        } = render(<ExportPage />);

        await waitFor(() => expect(getByText("April")));

        const regIds = await store.timesheets.addRegistrationsAsync(registrations);

        await waitFor(() => expect(getByText("Foobar 5")));

        let items = container.querySelectorAll(".grouped-registration-header-date");
        expect(
            items.length
        ).toBe(3);

        expect(Array.from(items).map(el => (el as HTMLElement).textContent))
            .toEqual([
                "March 1st",
                "March 2nd",
                "March 24th",
            ]);

        await store.timesheets.deleteRegistrationsAsync(...regIds);

        await waitFor(() => expect(asFragment()).toMatchSnapshot());
    });
});
