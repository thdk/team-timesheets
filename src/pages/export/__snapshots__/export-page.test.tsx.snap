// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export Page should render without registrations 1`] = `
<DocumentFragment>
  <div
    class="flex  vertical"
  >
    <div
      class="flex  horizontal"
      style="align-items: center; justify-content: space-between; width: 100%;"
    >
      <div
        class="flex  horizontal"
        style="margin: 1em;"
      >
        <div
          class="first formField"
        >
          <div
            class="date-select-year mdc-select mdc-select--outlined"
            role="listbox"
          >
            <div
              class="mdc-select__anchor"
            >
              <i
                class="mdc-select__dropdown-icon"
              />
              <div
                aria-disabled="false"
                aria-haspopup="listbox"
                class="mdc-select__selected-text"
                role="button"
                tabindex="-1"
              >
                2020
              </div>
              <div
                class="mdc-notched-outline mdc-notched-outline--upgraded"
                notch="0"
              >
                <div
                  class="mdc-notched-outline__leading"
                />
                <div
                  class="mdc-notched-outline__notch"
                >
                  <span
                    class="mdc-floating-label mdc-floating-label--float-above"
                    for="select-Year"
                    style="transition-duration: 0s;"
                  >
                    Year
                  </span>
                </div>
                <div
                  class="mdc-notched-outline__trailing"
                />
              </div>
              <select
                class="rmwc-select__native-control date-select-year"
                tabindex="0"
              >
                <option
                  value="2020"
                >
                  2020
                </option>
                <option
                  value="2019"
                >
                  2019
                </option>
                <option
                  value="2018"
                >
                  2018
                </option>
                <option
                  value="2017"
                >
                  2017
                </option>
                <option
                  value="2016"
                >
                  2016
                </option>
              </select>
            </div>
          </div>
        </div>
        <div
          class="formField"
        >
          <div
            class="date-select-month mdc-select mdc-select--outlined"
            role="listbox"
          >
            <div
              class="mdc-select__anchor"
            >
              <i
                class="mdc-select__dropdown-icon"
              />
              <div
                aria-disabled="false"
                aria-haspopup="listbox"
                class="mdc-select__selected-text"
                role="button"
                tabindex="-1"
              >
                September
              </div>
              <div
                class="mdc-notched-outline mdc-notched-outline--upgraded"
                notch="0"
              >
                <div
                  class="mdc-notched-outline__leading"
                />
                <div
                  class="mdc-notched-outline__notch"
                >
                  <span
                    class="mdc-floating-label mdc-floating-label--float-above"
                    for="select-Month"
                    style="transition-duration: 0s;"
                  >
                    Month
                  </span>
                </div>
                <div
                  class="mdc-notched-outline__trailing"
                />
              </div>
              <select
                class="rmwc-select__native-control date-select-month"
                tabindex="0"
              >
                <option
                  value="0"
                >
                  January
                </option>
                <option
                  value="1"
                >
                  February
                </option>
                <option
                  value="2"
                >
                  March
                </option>
                <option
                  value="3"
                >
                  April
                </option>
                <option
                  value="4"
                >
                  May
                </option>
                <option
                  value="5"
                >
                  June
                </option>
                <option
                  value="6"
                >
                  July
                </option>
                <option
                  value="7"
                >
                  August
                </option>
                <option
                  value="8"
                >
                  September
                </option>
                <option
                  value="9"
                >
                  October
                </option>
                <option
                  value="10"
                >
                  November
                </option>
                <option
                  value="11"
                >
                  December
                </option>
              </select>
            </div>
          </div>
        </div>
      </div>
      <button
        class="mdc-button mdc-button--outlined"
        style="margin: 1em;"
      >
        <div
          class="mdc-button__ripple"
        />
        <span
          class="mdc-button__label"
        >
          Export
        </span>
      </button>
    </div>
    <div
      class="flex  vertical"
      style="padding-right: 1em; align-items: flex-end;"
    />
    <ul
      class="mdc-list"
      style="width: 100%;"
    >
      <li
        class="mdc-list-divider"
        role="separator"
      />
      <li
        class="mdc-list-item mdc-list-item--disabled"
        tabindex="0"
      >
        <span
          class="mdc-list-item__text"
        >
          Total in September
        </span>
        <div
          class="mdc-list-item__meta"
        >
          0 hours
        </div>
      </li>
      <li
        class="mdc-list-divider"
        role="separator"
      />
    </ul>
  </div>
</DocumentFragment>
`;
